lightning/src/util/persist.rs

```diff
@@ -194,3 +206,857 @@ where
+				KVStoreUpdatingPersisterError::StorageReadFailed {
+					reason: e,
+					context: key,
+				}
+			})?)
+			.map_err(|e| KVStoreUpdatingPersisterError::UpdateDecodeFailed {
+				reason: e,
+				context: key,
+			})?,
+		)
+	}
+
+	/// Delete updates with an update_id lower than the given channel monitor.
+	fn delete_stale_updates<ChannelSigner: WriteableEcdsaChannelSigner>(
+		&self,
+		channel_id: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+	) -> io::Result<()> {
+		let monitor_name: MonitorName = channel_id.into();
+		let update_names = self.list_update_names(&monitor_name)?;
+		for update_name in update_names {
+			let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+			if update.update_id != CLOSED_CHANNEL_UPDATE_ID
+				&& update.update_id <= monitor.get_latest_update_id()
+			{
+				let ns = self.monitor_update_namespace(&monitor_name);
+				let key = update_name.as_str();
+				self.kv.remove(&ns, key)?;
+			}
+		}
+		Ok(())
+	}
+
+	/// Read a monitor, discarding sentinel bytes if they are found.
+	fn discard_sentinel_bytes(
+		reader: &mut <<K as Deref>::Target as KVStore>::Reader,
+	) -> Result<Chain<Cursor<Vec<u8>>, BufReader<&mut <<K as Deref>::Target as KVStore>::Reader>>,io::Error> {
+		let mut bufreader = BufReader::new(reader);
+		let mut buffer = [0; KVSTORE_UPDATING_PERSISTER_MONITOR_PREPEND_SENTINEL.len()];
+
+		// "Peek" at the sentinel bytes by reading them into buffer
+		bufreader.read_exact(&mut buffer)?;
+
+		if buffer == *KVSTORE_UPDATING_PERSISTER_MONITOR_PREPEND_SENTINEL {
+			// This a monitor with updates possibly stored outside of it, made by this struct,
+			// so strip the sentinel bytes
+			let buffer_reader = Cursor::new(vec![]);
+			let chained_reader = buffer_reader.chain(bufreader);
+			Ok(chained_reader)
+		} else {
+			// This is a "regular" monitor, so chain the tested bytes to the reader
+			let buffer_reader = Cursor::new(buffer.to_vec());
+			let chained_reader = buffer_reader.chain(bufreader);
+			Ok(chained_reader)
+		}
+	}
+}
+
+impl<ChannelSigner: WriteableEcdsaChannelSigner, K: Deref, L: Deref> Persist<ChannelSigner>
+	for KVStoreUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	// TODO: We really need a way for the persister to inform the user that its time to crash/shut
+	// down once these start returning failure.
+	// A PermanentFailure implies we should probably just shut down the node since we're
+	// force-closing channels without even broadcasting!
+	fn persist_new_channel(
+		&self,
+		funding_txo: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		_update_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		let monitor_name = match MonitorName::try_from(funding_txo) {
+			Ok(n) => n,
+			Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+		};
+		let key = monitor_name.as_str();
+		let mut monitor_bytes = KVSTORE_UPDATING_PERSISTER_MONITOR_PREPEND_SENTINEL.to_vec();
```

Comment:
I don't think that will work? I need the sentinel bytes as well. I guess I could do that + sentinel.len(), and write the sentinel first, then the monitor. This is to avoid additional allocation from `extend`?

Reply:
Yep this was obviated once we switched to a byte vec.